src/thermo/CoverageDependentSurfPhase.cpp

```diff
@@ -0,0 +1,772 @@
+                double h_slope_low = item.enthalpy_params[0];
+                double h_slope_high = item.enthalpy_params[1];
+                double h_cov_change = item.enthalpy_params[2];
+                if (cov[j] <= h_cov_change) {
+                    m_h_cov[k] += h_slope_low * cov[j];
+                } else {
+                    m_h_cov[k] += h_slope_high
+                        * (cov[j] - h_cov_change)
+                        + (h_cov_change * h_slope_low);
+                }
+                double s_slope_low = item.entropy_params[0];
+                double s_slope_high = item.entropy_params[1];
+                double s_cov_change = item.entropy_params[2];
+                if (cov[j] <= s_cov_change) {
+                    m_s_cov[k] += s_slope_low * cov[j];
+                } else {
+                    m_s_cov[k] += s_slope_high
+                        * (cov[j] - s_cov_change)
+                        + (s_cov_change * s_slope_low);
+                }
+            }
+        }
+        
+        // For interpolative model
+        if (m_has_interpolative_dependency) {
+            for (auto& item : m_InterpolativeDependency) {
+                size_t k = speciesIndex(item.name_k);
+                size_t j = speciesIndex(item.name_j);
+                vector_fp h_covs = item.enthalpy_coverages;
+                size_t i_inter = 0;
+                for (size_t i = 0; i < (h_covs.size()-1); i++) {
+                    if (h_covs[i] <= cov[j] && cov[j] <= h_covs[i+1]) {
+                        i_inter = i;
+                        break;
+                    }
+                }
+                double h_left = item.enthalpies[i_inter];
+                double h_right = item.enthalpies[i_inter+1];
+                m_h_cov[k] += (h_right - h_left) / 
+                    (h_covs[i_inter+1] - h_covs[i_inter]) 
+                    * (cov[j]  - h_covs[i_inter])
+                    + h_left;
+
+                vector_fp s_covs = item.entropy_coverages;
+                i_inter = 0;
+                for (size_t i = 0; i < (s_covs.size()-1); i++) {
+                    if (s_covs[i] <= cov[j] && cov[j] <= s_covs[i+1]) {
+                        i_inter = i;
+                        break;
+                    }
+                }
+                double s_left = item.entropies[i_inter];
+                double s_right = item.entropies[i_inter+1];
+                m_s_cov[k] += (s_right - s_left) / 
+                    (s_covs[i_inter+1] - s_covs[i_inter]) 
+                    * (cov[j] - s_covs[i_inter])
+                    + s_left;
+            }
+        }
+    }
+
+    if (m_covstateNumlast != covstateNumnow || m_tlast != tnow || force) {
+        // For coverage-depedent heat capacity
+        if (m_has_heatcapacity_dependency) {
+            for (size_t k = 0; k < m_kk; k++) {
+            m_cp_cov[k] = 0.0;
+            }
+            vector_fp cov(m_kk, 0.0);
+            SurfPhase::getCoverages(cov.data());
```

Comment:
To avoid the inefficiency of repeatedly reallocating the `cov` array (every time themo properties are recalculated), I would suggest making `m_cov` a member of the class, and then you can just resize it as species are added, and call `getCoverages` once in this method. Also, I don't think there's a need to explicitly name the base class here when calling `getCoverages`.